### PR TITLE
Don't Fail LocalTaskJob on heartbeat

### PR DESCRIPTION
**Pull request by collinmcnulty**
The local task job heartbeat will crash the task if it fails. This is counter to the goal of the heartbeat, which is that it's ok if the task temporarily loses connection to the DB as long as it re-establishes it within the zombie timeout.

I've used an extremely broad exception here. I'm open to something more narrow if we have a smaller error type that will still cover the problems not controllable by the Airflow code, but I think it's reasonable to use a very broad exception.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/jobs/local_task_job_runner.py [hunk 1]
```diff
@@ -208,9 +208,16 @@ def sigusr2_debug_handler(signum, frame):
...
+                    try:
+                        perform_heartbeat(
+                            job=self.job, heartbeat_callback=self.heartbeat_callback, only_if_necessary=False
+                        )
+                    except Exception as e:
+                        # Failing the heartbeat should never kill the localtaskjob
+                        # If it repeatedly can't heartbeat, it will be marked as a zombie anyhow
+                        warn_string = f"Heartbeat failed with Exception: {e}"
+                        self.log.warning(warn_string)
+                        pass
```
Comment (RNHTTR) [inline airflow/jobs/local_task_job_runner.py]: No need for pass at this point

File: airflow/jobs/local_task_job_runner.py [hunk 2]
```diff
@@ -208,9 +208,16 @@ def sigusr2_debug_handler(signum, frame):
...
-                        job=self.job, heartbeat_callback=self.heartbeat_callback, only_if_necessary=False
-                    )
+                    try:
+                        perform_heartbeat(
+                            job=self.job, heartbeat_callback=self.heartbeat_callback, only_if_necessary=False
+                        )
+                    except Exception as e:
+                        # Failing the heartbeat should never kill the localtaskjob
+                        # If it repeatedly can't heartbeat, it will be marked as a zombie anyhow
+                        warn_string = f"Heartbeat failed with Exception: {e}"
```
Comment (RNHTTR) [inline airflow/jobs/local_task_job_runner.py]: Logging shouldn't use f-strings https://docs.astral.sh/ruff/rules/logging-f-string/#:~:text=Using%20f%2Dstrings%20to%20format,argument%20defers%20formatting%20until%20required.

**Review by RNHTTR — CHANGES_REQUESTED** [review]

File: airflow/jobs/local_task_job_runner.py (hunk 2 above)
Comment (collinmcnulty) [reply airflow/jobs/local_task_job_runner.py]: fixed

File: airflow/jobs/local_task_job_runner.py (hunk 1 above)
Comment (collinmcnulty) [reply airflow/jobs/local_task_job_runner.py]: fixed

**Comment by collinmcnulty** [comment]
Sorry Jed and Jarek for the incorrect notification, I messed up the rebase process and included extra changes. Now fixed.

**Review by jscheffl — APPROVED** [review]

**Review by RNHTTR — APPROVED** [review]

**Comment by jedcunningham** [comment]
Cherry-pick pr: https://github.com/apache/airflow/pull/41810

**Review by tthompson691 — COMMENTED** [review]
This is a long overdue improvement!

**Pull request merged**